pyroomacoustics/open_sofa_interpolate.py

```diff
@@ -0,0 +1,556 @@
+        : Parameter
+        ------------------------------
+
+        azimuth:nd.array() (int)  [-pi,pi]
+        colatitude:nd.array() (int) [0,pi]
+
+        : Return : nd.array() return nearest index on the grid based on the input angles using kDtree
+        """
+
+        longitude=azimuth.reshape(-1,1)
+        latitude=colatitude.reshape(-1,1)
+
+        start=timer()
+        if self.interpolate:
+            dd,ii=self.nn_kd_tree_rotated_fibo_grid.query(np.hstack((longitude,latitude)))
+        else:
+            dd, ii = self.nn_kd_tree_rotated_sofa_grid.query(np.hstack((longitude,latitude)))  # Query on the rotated set of points
+        end=timer()
+        print("Time taken For Query Once ", end - start)
+        return ii
+
+    def neareast_neighbour(self,index):
+        """
+        Interpolated or non interpolated response given the azimuth and colatitude of the angle of arrival (microphone) / angle of departure (source)
+        Query on a KDTree given azimuth and elevation and find the most neareast point according to that on the sphere.
+
+        :Parameter
+        -------------------------
+        azimuth: (int)  [-pi,pi]
+        colatitude: (int) [0,pi]
+
+        :Return (np.array) [Coeffs of DFT]
+        -------------------------
+        FIR in frequency resolution for the particular query angle.
+
+        """
+
+        if self.interpolate:
+            #dd, ii = self.nn_kd_tree_rotated_fibo_grid.query([longitude, latitude])  #Query on the rotated set of points
+            return self.sh_coeffs_expanded_target_grid[index, :]
+        else:
+            #dd, ii = self.nn_kd_tree_rotated_sofa_grid.query([longitude, latitude])  # Query on the rotated set of points
+            return self.sofa_msr_fir[index, :]
+
+    def change_orientation(self, azimuth_change, colatitude_change):
+        """
+        Change of rotation without repeating the whole process only works for the sofa file where we are doing interpolation on fibo sphere
+        Made for IWAENC.
+
+        """
+
+        n_c = (colatitude_change)
+        n_a = (azimuth_change)
+        R_y = np.array([[np.cos(n_c), 0, np.sin(n_c)], [0, 1, 0], [-np.sin(n_c), 0, np.cos(n_c)]])
+        R_z = np.array([[np.cos(n_a), -np.sin(n_a), 0], [np.sin(n_a), np.cos(n_a), 0], [0, 0, 1]])
+        res = np.matmul(R_z, R_y)
+
+        self.rotated_fibo_points = np.matmul(res, self.points.T)
+
+        self.rotated_fibo_phi, self.rotated_fibo_theta, self.rotated_fibo_r = cart2sphere(self.rotated_fibo_points.T)
+
+        self.nn_kd_tree_rotated_fibo_grid = cKDTree(
+            np.hstack((self.rotated_fibo_phi.reshape(-1, 1), self.rotated_fibo_theta.reshape(-1, 1))))
+
+
+
+
+
+
+    def plot(self,freq_bin=0):
+        """
+        Plot 3D plots of the SOFA file , interpolated fibonnacci sphere and roatated fibo and sofa spheres.
+
+        Display gain for a specified freq bin
+
+
+        :Parameter (int) freq_bin:Bin number
+
+        """
+
```

Comment:
To protect matplotlib import, one solution is to import inside the function. Or use a `has_matplotlib` flag set when doing the initial import.